src/client/auth/oidc.rs

```diff
@@ -69,33 +143,77 @@ pub struct IdpServerInfo {
     pub request_scopes: Option<Vec<String>>,
 }
 
+#[derive(Debug)]
+pub struct CallbackContext {
+    pub timeout_seconds: Option<Instant>,
+    pub version: i32,
+    pub refresh_token: Option<String>,
+    pub idp_info: Option<IdpServerInfo>,
+}
+
 #[derive(TypedBuilder)]
 #[builder(field_defaults(setter(into)))]
 #[non_exhaustive]
 pub struct IdpServerResponse {
     pub access_token: String,
-    pub expires: Option<Instant>,
+    pub expires_in_seconds: Option<Instant>,
     pub refresh_token: Option<String>,
 }
 
-#[derive(Debug)]
-#[non_exhaustive]
-pub struct RequestParameters {
-    pub deadline: Instant,
-}
-
 pub(crate) async fn authenticate_stream(
     conn: &mut Connection,
     credential: &Credential,
     server_api: Option<&ServerApi>,
 ) -> Result<()> {
-    let source = credential.source.as_deref().unwrap_or("$external");
-    let callbacks = credential
-        .oidc_callbacks
+    // RUST-1662: Attempt speculative auth first, only works with a cache.
+    // First handle speculative authentication. If that succeeds, we are done.
+
+    match credential
+        .oidc_callback
         .as_ref()
         .ok_or_else(|| auth_error("no callbacks supplied"))?
-        .clone();
+        .callback
+        .clone()
+    {
+        Callback::Machine(callback) => {
+            authenticate_machine(conn, credential, server_api, callback).await
+        }
+        Callback::Human(callback) => {
+            authenticate_human(conn, credential, server_api, callback).await
+        }
+    }
+}
+
+async fn update_oidc_cache(
+    credential: &Credential,
+    response: &IdpServerResponse,
+    token_gen_id: i32,
+) {
+    {
+        let mut cache = credential
+            .oidc_callback
+            .as_ref()
+            // unwrap() is safe here because authenticate_human is only called if oidc_callback is Some
+            .unwrap()
+            .cache
+            .write()
+            .await;
+        cache.access_token = Some(response.access_token.clone());
+        cache.refresh_token = response.refresh_token.clone();
+        cache.last_call_time = Instant::now();
+        cache.token_gen_id = token_gen_id;
+    }
+}
 
+async fn authenticate_human(
```

Comment:
It does look like `authenticate_human` and `authenticate_machine` are largely the same, so it seems like having a `CallbackKind` that gets passed in here and having explicit `if` blocks would reduce the code duplication.